ostd/src/orpc/oqueue/ringbuffer/mpmc.rs

```diff
@@ -311,8 +296,19 @@ impl<T, const STRONG_OBSERVERS: bool, const WEAK_OBSERVERS: bool>
                 {
                     // We have exclusive write access to the slot, safe to write.
                     unsafe { slot.store(data) };
-                    // Mark the slot as writeable
-                    slot.turn.store(self.turn(head, true), Ordering::Release);
+                    if self.tail.load(Ordering::Acquire) > head
+                        && self
+                            .strong_observer_tails
+                            .iter()
+                            .all(|t| t.load(Ordering::Acquire) > head)
+                    {
+                        // Mark the slot as writable so that the produce isn't blocked
+                        slot.turn
+                            .store(self.next_turn(head, false), Ordering::Release);
+                    } else {
+                        // Mark the slot as writeable
+                        slot.turn.store(self.turn(head, true), Ordering::Release);
+                    }
```

Comment:
This appears to be identical to the code above. I think having `store_slot` method might be appropriate. It should have well defined semantics: Write the value into the already allocated slot and update the turn based on the presence or absence of observers and consumers.